File: String/22generateParentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans = []
        
        def check(s = '', left = 0, right = 0):
            if len(s) == 2*n:
                ans.append(s)
                return
            if left < n:
                check(s+'(',left+1,right)
            if right < left:
                check(s+')',left,right+1)
        check()         
        return ans

File: String/22GenerateParentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> rev;
        string res = "";
        helper(n, 0, 0, res, rev);
        return rev;
    }
    void helper(int n, int l, int r, string res, vector<string>& rev){
        if(l==n && r==n) rev.push_back(res);
        if(l<n) helper(n, l+1, r, res+'(', rev);
        if(r<l) helper(n, l, r+1, res+')', rev);
    }
};